sdk/dotnet/Pulumi/Resources/CustomResource.cs

```diff
@@ -18,20 +18,49 @@ public class CustomResource : Resource
         /// </summary>
         // Set using reflection, so we silence the NRT warnings with `null!`.
         [Output(Constants.IdPropertyName)]
-        public Output<string> Id { get; private set; } = null!;
+        public Output<string> Id { get; private protected set; } = null!;
 
         /// <summary>
-        /// Creates and registers a new managed resource.  t is the fully qualified type token and
-        /// name is the "name" part to use in creating a stable and globally unique URN for the
-        /// object. dependsOn is an optional list of other resources that this resource depends on,
-        /// controlling the order in which we perform resource operations.Creating an instance does
-        /// not necessarily perform a create on the physical entity which it represents, and
-        /// instead, this is dependent upon the diffing of the new goal state compared to the
-        /// current known resource state.
+        /// Creates and registers a new managed resource.  <paramref name="type"/> is the fully
+        /// qualified type token and <paramref name="name"/> is the "name" part to use in creating a
+        /// stable and globally unique URN for the object.  <see cref="ResourceOptions.DependsOn"/>
+        /// is an optional list of other resources that this resource depends on, controlling the
+        /// order in which we perform resource operations.  Creating an instance does not necessarily
+        /// perform a create on the physical entity which it represents, and instead, this is
+        /// dependent upon the diffing of the new goal state compared to the current known resource
+        /// state.
         /// </summary>
+        /// <param name="type">The type of the resource.</param>
+        /// <param name="name">The unique name of the resource.</param>
+        /// <param name="args">The arguments to use to populate the new resource.</param>
+        /// <param name="options">A bag of options that control this resource's behavior.</param>
 #pragma warning disable RS0022 // Constructor make noninheritable base class inheritable
-        public CustomResource(string type, string name, ResourceArgs? args, CustomResourceOptions? options = null)
-            : base(type, name, custom: true, args ?? ResourceArgs.Empty, options ?? new CustomResourceOptions())
+        public CustomResource(string type, string name, ResourceArgs? args, CustomResourceOptions? options)
```

Comment:
Consider an existing resource:

```c#
public class MyResource : CustomResource
{
    public MyResource()
        : base("type", "name", ResourceArgs.Empty)
    {
    }
}
```

If I left `= null` here, the above code would now see a new compiler error:

> The call is ambiguous between the following methods or properties: 'CustomResource.CustomResource(string, string, ResourceArgs?, CustomResourceOptions?)' and 'CustomResource.CustomResource(string, string, ResourceArgs?, CustomResourceOptions?, bool)'

The approach I took when adding the new optional parameters is from Jon Skeet ("Library version 1.1c"): https://codeblog.jonskeet.uk/2018/04/13/backward-compatibility-and-overloading/

> ## Optional parameters
>
> Optional parameters bring their own kind of pain. Suppose we have one optional parameter, but wish to add a second. We have three options, shown as 1.1a, 1.1b and 1.1c below.
>
> ```c#
> // Library version 1.0
> public void Foo(string x = "")
> 
> // Library version 1.1a
> // Keep the existing method, but add another one with two optional parameters.
> public void Foo(string x = "")
> public void Foo(string x = "", string y = "")
> 
> // Library version 1.1b
> // Just add the parameter to the existing method.
> public void Foo(string x = "", string y = "")
> 
> // Library version 1.1c
> // Keep the old method but make the parameter required, and add a new method
> // with both parameters optional.
> public void Foo(string x)
> public void Foo(string x = "", string y = "")
> ```
>
> Let’s think about a client that makes two calls:
>
> ```c#
> // Client
> static void Method()
> {
>     var library = new Library();
>     library.Foo();
>     library.Foo("xyz");
> }
> ```
>
> Library 1.1a keeps binary compatiblity, but breaks source compatibility: the `library.Foo()` is now ambiguous. The C# overloading rules prefer a method that doesn’t need the compiler to “fill in” any optional parameters, but it doesn’t have any preference in terms of _how many_ optional parameters are filled in.
>
> Library 1.1b keeps source compatibility, but breaks binary compatibility. Existing compiled code will expect to call a method with a single parameter – and that method no longer exists.
>
> Library 1.1c keeps binary compatibility, but is potentially odd around source compatibility. The `library.Foo()` call now resolves to the two-parameter method, whereas `library.Foo("xyz")` resolves to the one-parameter method (which the compiler prefers over the two-parameter method because it doesn’t need to fill in any optional parameters). That may very well be okay, if the one-parameter version simply delegates to the two-parameter version using the same default value. It feels odd for the meaning of the first call to change though, when the method it _used_ to resolve to still exists.

Reply:
Interesting... I thought 1.1c breaks binary compatibility. I guess Jon Skeet knows better - it mostly makes sense now.